lib/ddtrace/life_cycle.rb

```diff
@@ -0,0 +1,56 @@
+require 'ddtrace/tracing/runtime'
+
+module Datadog
+  # Manages the {Runtime} life cycle for the host application.
+  #
+  # The {Runtime} is a container of all stateful tracer internals.
+  #
+  # {LifeCycle} ensure that the {Runtime} is created,
+  # modified and decommissioned correctly during the host
+  # application's life cycle.
+  module LifeCycle
+    extend Forwardable
+
+    def_delegators \
+        :runtime,
+        :configure, :shutdown!,
+        :configuration,
+        :health_metrics, :logger, :profiler, :runtime_metrics, :tracer
+
+    protected
+
+    def start!(runtime = Tracing::Runtime.new)
+      @runtime = runtime
+      runtime.startup!
+    end
+
+    private
+
+    attr_reader :runtime
+
+    # Only used internally for testing
+    def started?
+      !@runtime.nil?
+    end
+
+    # Only used internally for testing
+    def tear_down!
+      shutdown! if started?
+
+      # TODO: this is actually part of contrib
+      # TODO: should probably be moved to inside contrib/
+      # Reset stateful registry data
+      registry.each do |data|
+        data.klass.reset_configuration!
+      end
+
+      @runtime = nil
+    end
+
+    # Only used internally for testing
+    def restart!
+      tear_down!
+      start!
+    end
```

Comment:
If these methods are only used for testing, can we add them here from somewhere inside the `spec/` folder? That way we would entirely avoid misuse.

Reply:
I think I tried that, but they happen too early in the application `require` cycle.
I'll give it another shot at it.